### ISSUE TITLE
ListBox does not trigger "activate" on mouse click or double-click

### ISSUE THREAD

**Issue by joris-r** [gtk-rs/gtk#520]
I'm trying to use `connect_activate` on a `ListBoxRow`. It's working with a keyboard event (the return key) but not with a mouse click.

See the code below. I'm expecting to see the "activation ..." println on a click on the widget.
What do you think ?

    extern crate gtk;
    use gtk::prelude::*;

    fn main() {
        if gtk::init().is_err() {
            println!("Failed to initialize GTK.");
            return;
        }

        let window = gtk::Window::new(gtk::WindowType::Toplevel);
        window.set_size_request(640,480);
        
        window.set_title("test");
        window.connect_delete_event(|_, _| {
            gtk::main_quit();
            Inhibit(false)
        });
            
        let lb = gtk::ListBox::new();
        window.add(&lb);
        
        assert!(lb.get_activate_on_single_click() == true);
            
        for s in ["one", "two", "three"].iter() {
            let row = gtk::ListBoxRow::new();
            lb.add(&row);
            row.add(&gtk::Label::new(s as &str));
            
            assert!(row.get_activatable() == true);
            
            let s = s.clone();
            row.connect_activate(move |_| {
                println!("activation {}", s)
            });
            
        }
        
        window.show_all();
        gtk::main();
    }

**Comment by GuillaumeGomez** [comment]
Take a look here: https://github.com/GuillaumeGomez/process-viewer/blob/master/src/procs.rs#L82-L96

**Comment by joris-r** [comment]
Thanks but `ListBox` doesn't have a model (unlike tree_view), so with a signal like `connect_row_selected` I can only call `ListBoxRow::get_index`. It's not very convenient and as the signal for selected is on the `ListBox` and not on the `ListBoxRow` I would have to manually manage the link between the different row and my corresponding data.

That would be better to attach a closure to the row but there is only the signal `connect_activate` available on a `ListBoxRow`.

**Comment by EPashkin** [comment]
There also `connect_row_activated` that works with mouse click in your example

**Comment by GuillaumeGomez** [comment]
There is this function too: https://developer.gnome.org/gtk3/stable/GtkListBox.html#gtk-list-box-set-activate-on-single-click

**Comment by joris-r** [comment]
@GuillaumeGomez  Yes I saw it, and in my example there is a `assert!(lb.get_activate_on_single_click() == true);`. Anyway activation should occurs on double-click and it doesn't work either.

@EPashkin thanks I'm testing that: if, in my example, I add this before the loop

    lb.connect_row_activated(move |_,lbr| {
        println!("youyou ?");
        assert!(lbr.get_activatable() == true);
    });

I see the the activation with a single click. So the question is: why is the activation not transferred from the ListBox to the ListBoxRow?
I have tried to call myself `lbr.activate()` but this provoke a loop (and so a stack overflow).

It looks like a bug, doesn't it ?

**Comment by EPashkin** [comment]
Maybe Label that in ListBoxRow don't pass click to row.

**Comment by joris-r** [comment]
No the child relation ship is the other way: ListBox --> ListBoxRow --> Label

**Comment by joris-r** [comment]
And now I have applied the workaround on my actual project. It's working but I see the following in the terminal

    (.:12069): Gtk-CRITICAL **: gtk_widget_grab_focus: assertion 'GTK_IS_WIDGET (widget)' failed
    (.:12069): Gtk-CRITICAL **: gtk_widget_queue_draw: assertion 'GTK_IS_WIDGET (widget)' failed
    (.:12069): Gtk-CRITICAL **: gtk_widget_get_accessible: assertion 'GTK_IS_WIDGET (widget)' failed
    (.:12069): Gtk-CRITICAL **: gtk_list_box_row_get_activatable: assertion 'GTK_IS_LIST_BOX_ROW (row)' failed

Look likes there is something wrong internally

**Comment by EPashkin** [comment]
Or you just make typo ;)

**Comment by EPashkin** [comment]
Don't know about Gtk but normally mouse click go from child to parent if child don't process it.

**Comment by joris-r** [comment]
Yeah of course, I'm not an expert on gtk at all ;)
But the documentation says 

> Although a ListBox must have only ListBoxRow children you can add any kind of widget to it via ContainerExt::add, and a ListBoxRow widget will automatically be inserted between the list and the widget.

So I think I (as a library simple user) should not be able to mess up this.

**Comment by SamWhited** [comment]
I am also running into this issue and cannot use the `listbox.connect_row_activated` because in the callback I need information that is only available when the `ListBoxRow` is added.

**Comment by tmiasko** [comment]
"active" signal on `ListBoxRow` is an action signal that can be used to
activate the row by emitting it. One can connect to it, but there is no
guarantee that all row activation will be performed through this signal.

The "row-selected" or "selected-rows-changed" is the one you should connect to.

**Comment by mnussbaum** [comment]
I'm running into this as well. Using `connect_activated` on `ListBoxRow` ensures that as the data presented by the `ListBox` changes only callbacks for the active data set will be invoked.

AFAICT `ListBox` doesn't provide a method to get the current selection like tree views do. So attempting to work-around this problem by using a method like `connect_row_activated` on the `ListBox` means I need clone the data set presented in the list box so that I can index into the clone in my activated callback function.

Referencing a clone in the callback is problematic because as the data shown in the list box changes my callbacks aren't updated, and they start to act on stale data.

I'm a GTK noob, so maybe I'm missing something about how to get the row data out of the arguments passed to the `connect_row_activated` callback?

**Comment by sdroege** [comment]
Instead of actually copying the data, you could store your data in an `Rc` so that the same data can be accessed from multiple places at once safely.
